packages/commonwealth/client/scripts/state/api/threads/helpers/cache.ts

```diff
@@ -0,0 +1,254 @@
+            if (foundThreadIndex > -1) {
+                updatedThreads[foundThreadIndex] = {
+                    ...updatedThreads[foundThreadIndex],
+                    topic: newTopic,
+                }
+            }
+            queryClient.setQueryData(k, () => updatedThreads);
+        }
+
+        // 2- for bulk queries - filter the existing thread that has old topic and refetch queries for the updates
+        // topic id ideally we should not refetch - TODO: find a way to make this consistent
+        if (k[2] === cacheTypes.BULK_THREADS) {
+            // filter from old topic query
+            if (k[3] === oldTopicId || k[3] === undefined) {
+                const existingData: IExistingThreadState = queryClient.getQueryData(k)
+                const pages = [...(existingData.pages || [])]
+                let foundThreadIndex = -1
+                const foundPageIndex = pages.findIndex(p => {
+                    const index = p.data.threads.findIndex(t => t.id === threadId)
+                    if (index > -1) foundThreadIndex = index
+                    return index > -1 ? true : -1
+                })
+
+                if (foundPageIndex > -1 && foundThreadIndex > -1) {
+                    pages[foundPageIndex].data.threads = pages[foundPageIndex].data.threads.filter(x => x.id !== threadId)
+                }
+
+                return {
+                    ...existingData,
+                    pages
+                }
+            }
+            // and refetch new topic queries
+            if (k[3] === newTopic.id || k[3] === undefined) {
+                queryClient.cancelQueries(k)
+                queryClient.refetchQueries(k)
+            }
+        }
+    })
+
+}
+
+const addThreadInAllCaches = (chainId: string, newThread: Thread) => {
+    // refetch all caches for the thread topic and also the general cache
+    const queryCache = queryClient.getQueryCache()
+    const queryKeys = queryCache.getAll().map(cache => cache.queryKey)
+    const keysForThreads = queryKeys.filter(x => x[0] === ApiEndpoints.FETCH_THREADS && x[1] === chainId)
+
+    keysForThreads.map((k) => {
+        // TODO: this is improper, we are essentially clearing cache when a thread is added. This is done to ensure
+        // we have the correct thread ordering when refetching threads, but ideally we should find a way to correctly
+        // position the thread in cache
+        if (
+            (k[2] === cacheTypes.BULK_THREADS && (k[3] === newThread.topic.id || k[3] === undefined)) ||
+            k[2] === cacheTypes.ACTIVE_THREADS
+        ) {
+            queryClient.cancelQueries(k)
+            queryClient.refetchQueries(k)
+        }
+        // TODO: for now single cache will fetch the thread - not adding its state, ideally we should
+        // add the thread here
+    })
+}
+
+export { addThreadInAllCaches, removeThreadFromAllCaches, updateThreadInAllCaches, updateThreadTopicInAllCaches };
+
```

Comment:
while the idea for this file is great, I am scared of it to be honest. I would say it is super fragile and bug prone. Imagine if one little thing like property or anything will change in the threads model -- probably we are screw here then. Also debugging it might be tough in case of any issues (and they will happen for sure after such a big refactor).

I can see two options here:
1. we keep this file but we have to make it much more readable. 
  - avoid long and complicated `conditional expressions` => the code inside `if ()`. extract it to the meaningful variable and use it
  - avoid nested if statements
  - unit tests
  - proper descriptions WHY this step is necessary

2. instead of manipulating cache we can invalidate query. Yes, the downside of it is that we are doing the API call but we if this call is not expensive/slow, I would definitely opt for this solution. If we use react-query, it does not mean that we have to handle EVERYTHING with cache. This library should make application more performant (and already does as we can notice) but it should not make our life more complicated. This kind of the magic-cache-updater makes me think that we will suffer from unknown behaviours. Please notice that most of the api calls are about fetching threads. Minority is creating/editing/deleting so I really believe that we won't suffer from extra api call in this case (when using invalidate query). 

cc @jnaviask

Reply:
> Please notice that most of the api calls are about fetching threads. Minority is creating/editing/deleting so I really believe that we won't suffer from extra api call in this case (when using invalidate query).

I kind of disagree here, if we invalidate then by extension we are re-fetching and the only place we will be invalidating will be in mutations - so every mutation will wipe out the data. For context, we have ~5 fetches for threads and ~15 mutations.

Reply:
In my opinion we could do the cache manipulation but only if the "[option A](https://github.com/hicommonwealth/commonwealth/pull/4588#discussion_r1282080806)" is already developed. Not other way round. 

@mzparacha by saying this 
> Please notice that most of the api calls are about fetching threads. Minority is creating/editing/deleting
I did not mean the number of files/actions we have defined in the codebase. I meant the number of actual api calls that are performed by the users (this is my ballpark but very possible imo). 

So I think we should go with invalidation pattern for now which is explained here:
- https://tanstack.com/query/v3/docs/react/guides/query-invalidation
- https://tanstack.com/query/v3/docs/react/guides/invalidations-from-mutations

So for example if you edit thread, you invalidate threads. If you delete the thread, you invalidate threads etc. `invalidateQueries` is very flexible so we can granulate the invalidation to make sure eg that the threads from other communities are not wiped out. I believe that those couple of single lines `queryClient.invalidateQueries()` might replace the whole `cache.ts` file and its dangers.

When the "option A" is ready we can follow this https://tanstack.com/query/v3/docs/react/guides/updates-from-mutation-responses so we will save extra api call.

Reply:
Thanks for the feedback, I will add comments. And I already included myself in [this](https://github.com/hicommonwealth/commonwealth/issues/4664#issuecomment-1665676783) ticket, plz lmk if I should create a new one instead